File: tests/entity_number_indexer_test.py
import unittest
import kgekit.io

class EntityNumberIndexerTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        t, _ = kgekit.io.read_triples('tests/fixtures/triple.txt', "hrt", ' ')
        cls._indexer = kgekit.EntityNumberIndexer(t, "hrt")

    def test_attrs(self):
        self.assertEqual(self._indexer.entityIdMap(), {'/m/entity1': 0, '/m/entity2': 1, '/m/entity3': 2})
        self.assertEqual(self._indexer.relationIdMap(), {'/country': 1, '/produced_by': 0, '/produced_in': 2})
        self.assertEqual(self._indexer.indexes(), [kgekit.TripleIndex([0, 0, 1]), kgekit.TripleIndex([1, 1, 2]), kgekit.TripleIndex([0, 2, 2])])
        self.assertEqual(self._indexer.entities(), ['/m/entity1', '/m/entity2', '/m/entity3'])
        self.assertEqual(self._indexer.relations(), ['/produced_by', '/country', '/produced_in'])

if __name__ == '__main__':
    unittest.main()

File: src/kgekit/entity_number_indexer_test.cpp
#include "entity_number_indexer.h"

#include "test_helper.h"
#include <pybind11/pybind11.h>

namespace kgekit
{

using namespace Catch;
using namespace std;

TEST_CASE_METHOD(IndexerTestsFixture, "build index lazyly", "[indexing]")
{
    auto indexer = getIndexer<EntityNumberIndexer>();
    REQUIRE(indexer->isIndexBuilt() == false);
    indexer->getEntityIdMap();
    REQUIRE(indexer->isIndexBuilt() == true);
}

TEST_CASE_METHOD(IndexerTestsFixture, "returns indexes", "[indexing]")
{
    auto indexer = getIndexer<EntityNumberIndexer>();

    TripleIndex first(0, 0, 1);
    TripleIndex second(1, 1, 2);
    TripleIndex third(0, 2, 2);
    REQUIRE_THAT(indexer->getIndexes(),
                 Predicate<decltype(indexer->getIndexes())>(
                     [&](const auto& vec) -> bool {
                         return vec->at(0) == first &&
                            vec->at(1) == second &&
                            vec->at(2) == third;
                     }));
    REQUIRE_THAT(indexer->getEntityIdMap(),
                 Predicate<decltype(indexer->getEntityIdMap())>(
                     [](const auto& maps) -> bool {
                         return maps->size() == 3 &&
                             maps->at("/m/entity1") == 0 &&
                             maps->at("/m/entity2") == 1 &&
                             maps->at("/m/entity3") == 2; },
                     "entities were inserted."));
    REQUIRE_THAT(indexer->getRelationIdMap(),
        Predicate<decltype(indexer->getRelationIdMap())>(
            [](const auto& maps) -> bool {
                return maps->size() == 3 &&
                    maps->at("/produced_by") == 0 &&
                    maps->at("/country") == 1 &&
                    maps->at("/produced_in") == 2; },
            "relations were inserted."));
}

} // namespace kgekit
